### PR TITLE
Move caret/highlight painting to custom painters

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

This PR introduces 2 new fields `forgroundPainter` and `painter` to `RenderEditable`, and moves caret/text highlight painting logic into these painters, without introducing actual behavior changes (hopefully). 
https://docs.google.com/document/d/10YiBZ_5oj6erdbgPSgpvEzNZUa7nEMYaa0X2Tw1GAWo/edit#

## Tests

I added the following tests:

A few tests for custom painters.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1125,7 +1247,7 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     if (_selection == value)
       return;
     _selection = value;
-    _selectionRects = null;
+    _selectionPainter.highlightedRange = value;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:1250]: technically this is not needed, since the selection painter will repaint with the `RenderEditable`.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 2]
```diff
@@ -2266,32 +2248,33 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
-        baseOffset: _promptRectRange!.start,
-        extentOffset: _promptRectRange!.end,
-      ),
-    );
-
-    for (final TextBox box in boxes) {
-      canvas.drawRect(box.toRect().shift(effectiveOffset), _promptRectPaint);
+    _floatingCursorOn = state != FloatingCursorDragState.End;
+    _resetFloatingCursorAnimationValue = resetLerpValue;
+    if (_floatingCursorOn) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:2267]: This part is different. The caret calculation was originally partially deferred to `paint` and is now performed in `setFloatingCursor`.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+///   * The [notifyListeners] method is called, which typically happens when the
+///     painter's attributes change.
+///
+/// See also:
+///  * [RenderEditable.setForegroundPainter], which takes a [RenderEditablePainter]
+///    and sets it as the foreground painter of the [RenderEditable].
+///  * [RenderEditable.setPainter], which takes a [RenderEditablePainter]
+///    and sets it as the foreground painter of the [RenderEditable].
+///  * [CustomPainter] a similar class which paints within a [RenderCustomPaint].
+abstract class RenderEditablePainter extends ChangeNotifier {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:2454]: The painter doesn't have `hitTest` or `semanticsBuilder` because these currently don't seem to have a use case.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:2454]: The `RenderEditable` property was removed because you can use the same painter instance in different render objects simultaneously when the tree finalizes (for example in a `Hero` widget we'd keep 2 copies of the subtree so in case there's a painter in the subtree it will be attached to 2 different `RenderEditable`s).

File: packages/flutter/lib/src/rendering/editable.dart [hunk 4]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  set shouldPaint(bool value) {
+    if (shouldPaint == value)
+      return;
+    _shouldPaint = value;
+    notifyListeners();
+  }
+
+  CaretChangedHandler caretPaintCallback;
+
+  bool showRegularCaret = false;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:2582]: Does not trigger a repaint because this is floating cursor stuff. Might need a bit of audit when we move the caret painting stuff to a different place.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 5]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+
+  final Paint caretPaint = Paint();
+  late final Paint floatingCursorPaint = Paint();
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:2585]: Uses a different painter because of a gotcha in TestRecordingCanvas.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 6]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+
+  @override
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable) {
+    if (!shouldPaint)
+      return;
+
+    assert(renderEditable != null);
+    final TextSelection? selection = renderEditable.selection;
+
+    if (selection == null || !selection.isCollapsed /* || !selection.isValid*/)
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/rendering/editable.dart:2692]: I thought we shouldn't paint the caret when selection is (-1, -1)?

File: packages/flutter/test/cupertino/text_field_test.dart [hunk 7]
```diff
@@ -2190,14 +2190,14 @@ void main() {
 
     expect(lastCharEndpoint.length, 1);
     // The last character is now on screen.
-    expect(lastCharEndpoint[0].point.dx, moreOrLessEquals(786.73, epsilon: 0.01));
+    expect(lastCharEndpoint[0].point.dx, moreOrLessEquals(786.73, epsilon: 2));
```
Comment (LongCatIsLooong) [inline packages/flutter/test/cupertino/text_field_test.dart:2193]: Not sure why but I had to change the epsilon here so these tests pass.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 8]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  set selectionWidthStyle(ui.BoxWidthStyle value) {
+    assert(value != null);
+    if (_selectionWidthStyle == value)
+      return;
+    _selectionWidthStyle = value;
+    notifyListeners();
+  }
+
+  @override
+  void paint(Canvas canvas, Size size, RenderEditable renderEditable) {
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:2536]: Could you explain your reasoning in taking TextRange as a parameter and passing RenderEditable into the paint method?  As opposed to maybe dealing with TextBoxes directly.  I don't disagree with the decision, I'm just curious how you designed this at a high level.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 9]
```diff
@@ -2407,3 +2370,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  bool shouldRepaint(RenderEditablePainter? oldDelegate) {
+    if (identical(oldDelegate, this))
+      return false;
+    if (oldDelegate == null)
+      return highlightColor != null && highlightedRange != null;
+    return oldDelegate is! _TextHighlightPainter
+        || oldDelegate.highlightColor != highlightColor
+        || oldDelegate.highlightedRange != highlightedRange
+        || oldDelegate.selectionHeightStyle != selectionHeightStyle
+        || oldDelegate.selectionWidthStyle != selectionWidthStyle;
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/editable.dart:2564]: Adding on to my previous comment about taking a RenderEditable in `paint`, I think that has some implications here too.  If the layout of the RenderEditable changes, then we do need to repaint the text highlight as well, even if this shouldRepaint method returns false.  I guess that's something that's kept track of separately, though.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 5 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:2585]: How is this `late` but it still has an initial value?  I think I just have never encountered this NNBD situation.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 6 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:2692]: I remember this came up awhile back in https://github.com/flutter/engine/pull/19785, was there somewhere else that we were talking about this?

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 7 above)
Comment (justinmc) [reply packages/flutter/test/cupertino/text_field_test.dart:2193]: Weird, nothing jumps out at me as a cause.  It is a big epsilon, but if no golden tests are failing then maybe it's not a concern...

File: packages/flutter/test/rendering/editable_test.dart [hunk 10]
```diff
@@ -134,10 +134,12 @@ void main() {
         offset: 0,
       ),
     );
-    editable.layout(BoxConstraints.loose(const Size(1000.0, 1000.0)));
+    layout(editable, constraints: BoxConstraints.loose(const Size(500.0, 500.0)));
```
Comment (justinmc) [inline packages/flutter/test/rendering/editable_test.dart:137]: Why did you change this size?  I admit I don't think I've looked at these tests before.

File: dev/devicelab/bin/tasks/fullscreen_textfield_perf.dart [hunk 11]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (justinmc) [reply dev/devicelab/bin/tasks/fullscreen_textfield_perf.dart:1]: No opinion from me, I haven't done much with these devicelab perf tests before 🤷

**Review by justinmc — COMMENTED** [review]
A bunch of questions and a few small changes.  RenderEditable really needed to be broken up more and this looks like a great improvement along those lines.

I think the biggest question I have with the overall PR is about exposing this to users.  I see the power in being able to customize these painters, but are users asking for any specific features here, and what's the main expected use case?  I'm talking about accepting the two new parameters in RenderEditable and making RenderEditablePainter public.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 12]
```diff
@@ -1249,45 +1368,11 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
   /// Defaults to a padding with left, top and right set to 4, bottom to 5.
-  EdgeInsets get floatingCursorAddedMargin => _floatingCursorAddedMargin;
-  EdgeInsets _floatingCursorAddedMargin;
-  set floatingCursorAddedMargin(EdgeInsets value) {
-    if (_floatingCursorAddedMargin == value)
-      return;
-    _floatingCursorAddedMargin = value;
-    markNeedsPaint();
-  }
+  EdgeInsets floatingCursorAddedMargin;
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1371]: Currently the floating cursor implementation does not handle some edge cases properly, changing `floatingCursorAddedMargin` is one of them. The floating cursor doesn't reposition itself when the text input resizes or moves. The visual update only happens when the user moves the touch point (and we get a text channel message).

File: packages/flutter/lib/src/rendering/editable.dart (hunk 5 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:2585]: I think this does lazy initialization. The instance variable won't be initialized until its first usage.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 6 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:2692]: Not that I remember. Is the expected behavior of `(-1, -1)` documented anywhere? I couldn't seem to find anything about it.

The `TextEditingValue` constructor uses `(-1, -1)` as the default selection range for years. https://github.com/flutter/flutter/blame/d35a9db6fe439b1efaddffb19cea540b99917d16/packages/flutter/lib/src/services/text_input.dart#L82-L90

File: dev/benchmarks/macrobenchmarks/lib/src/fullscreenTextField.dart [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  '🤗 🤔 🤭 🤫 🤥 😶 😐 😑 😬 🙄 😯 😦 😧 😮 😲 🥱 😴 🤤 😪 😵 '
+  '🤐 🥴 🤢 🤮 🤧 😷 🤒 🤕 🤑 🤠 😈 👿 👹 👺 🤡 💩 👻 💀 ☠️ 👽 '
+  '👾 🤖 🎃 😺 😸 😹 😻 😼 😽 🙀 😿 😾'
+  '️ 😣 😖 😫 😩 🥺 😢 😭 😤 😠 😡 🤬 🤯 😳 🥵 🥶 😱 😨 😰 😥 😓 '
+  '🤗 🤔 🤭 🤫 🤥 😶 😐 😑 😬 🙄 😯 😦 😧 😮 😲 🥱 😴 🤤 😪 😵 '
+  '🤐 🥴 🤢 🤮 🤧 😷 🤒 🤕 🤑 🤠 😈 👿 👹 👺 🤡 💩 👻 💀 ☠️ 👽 '
+  '👾 🤖 🎃 😺 😸 😹 😻 😼 😽 🙀 😿 😾'
+  '👋 🤚 🖐 ✋ 🖖 👌 🤌 🤏 ✌️ 🤞 🤟 🤘 🤙 👈 👉 👆 🖕 👇 ☝️ 👍 👎'
+  ' ✊ 👊 🤛 🤜 👏 🙌 👐 🤲 🤝 🙏 ✍️ 💅 🤳 💪 🦾 🦵 🦿 🦶 👣 👂 '
+  '🦻 👃 🫀 🫁 🧠 🦷 🦴 👀 👁 👅 👄 💋 🩸';
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply dev/benchmarks/macrobenchmarks/lib/src/fullscreenTextField.dart:44]: Don't think so. I'll put some in.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 14]
```diff
@@ -262,40 +265,139 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+    _autocorrectHighlightPainter.highlightedRange = promptRectRange;
+
+    _caretPainter.caretColor = cursorColor;
+    _caretPainter.cursorRadius = cursorRadius;
+    _caretPainter.cursorOffset = cursorOffset;
+    _caretPainter.backgroundCursorColor = backgroundCursorColor;
+
+    // Invalidate last caret rect cache.
+    if (onCaretChanged == null)
+      _lastCaretRect = null;
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:305]: Oh I just did a search & replace. Good catch :+1: .

File: packages/flutter/lib/src/rendering/editable.dart [hunk 15]
```diff
@@ -211,16 +212,18 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
     ui.BoxWidthStyle selectionWidthStyle = ui.BoxWidthStyle.tight,
     bool? enableInteractiveSelection,
-    EdgeInsets floatingCursorAddedMargin = const EdgeInsets.fromLTRB(4, 4, 4, 5),
+    this.floatingCursorAddedMargin = const EdgeInsets.fromLTRB(4, 4, 4, 5),
     TextRange? promptRectRange,
     Color? promptRectColor,
     Clip clipBehavior = Clip.hardEdge,
     required this.textSelectionDelegate,
+    RenderEditablePainter? painter,
+    RenderEditablePainter? foregroundPainter,
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:226]: `EdtiableText` does not currently pass in any painters. I'm planning to do that when we move the existing painters out.

File: packages/flutter/test/cupertino/text_field_test.dart (hunk 7 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/cupertino/text_field_test.dart]: I reverted the epsilon changes. Tests are passing.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 6 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:2688]: No, I haven't seen any good docs about it.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 16]
```diff
@@ -2046,6 +2137,21 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     }
   }
 
+  /// Computes the offset to apply to the given [sourceOffset] so it perfectly
+  /// snaps to physical pixels.
+  Offset _snapToPhysicalPixel(Offset sourceOffset) {
```
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:2138]: Ah I missed both of those, you're right 👍

File: packages/flutter/lib/src/rendering/editable.dart (hunk 9 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart:2560]: Ok that sounds good to me, it should always get repainted when needed.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I re-read your design doc and refreshed myself on the migration plan at the end.  So it seems that the plan is to move the RenderEditablePainters into their own files and then pass them into RenderEditable via the backgroundPainter and foregroundPainter params.  I'm on board with that, especially that it will simplify rendering/editable.dart, which is overly long and complicated right now.

Also, this doesn't seem to interfere with anything from my [text editing refactor plan](https://docs.google.com/document/d/1QaVIr1bbOWJGNyyNsB75sXgTDRMuxP268sLGTBNDut0/edit#heading=h.pub7jnop54q0).  If should actually make it easier to pare down RenderEditable even further.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/editable.dart:226]: When we do that, would we also remove/deprecate some of the existing arguments to RenderEditable?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 17]
```diff
@@ -262,40 +265,135 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+    }
+    _foregroundPainter = newPainter;
   }
+  /// Sets the [RenderEditablePainter] to use for painting above this
+  /// [RenderEditable]'s text content.
+  ///
+  /// The new [RenderEditablePainter] will replace the previously specified
+  /// foreground painter, and schedule a repaint if the new painter's
+  /// `shouldRepaint` method returns true.
+  void setForegroundPainter(RenderEditablePainter? newPainter) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:335]: Why not a regular getter/setter?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 18]
```diff
@@ -262,40 +265,135 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+    }
+    _painter = newPainter;
+  }
+  /// Sets the [RenderEditablePainter] to use for painting beneath this
+  /// [RenderEditable]'s text content.
+  ///
+  /// The new [RenderEditablePainter] will replace the previously specified
+  /// painter, and schedule a repaint if the new painter's `shouldRepaint`
+  /// method returns true.
+  void setPainter(RenderEditablePainter? newPainter) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:362]: same

File: packages/flutter/lib/src/rendering/editable.dart [hunk 19]
```diff
@@ -2407,3 +2366,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+/// An interface that paints within a [RenderEditable]'s bounds, above or
+/// beneath its text content.
+///
+/// This painter is typically used for painting auxiliary content that depends
+/// on text layout metrics (for instance, for painting carets and text highlight
+/// blocks). It can paint independently from its [RenderEditable], allowing it
+/// to repaint without triggering a repaint on the entire [RenderEditable] stack
+/// when only auxiliary content changes (e.g. a blinking cursor) are present. It
+/// will be scheduled to repaint when:
+///   * It's assigned to a new [RenderEditable] and the [shouldRepaint] method
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:2438]: nit: space before this line

File: packages/flutter/lib/src/rendering/editable.dart [hunk 20]
```diff
@@ -2407,3 +2366,407 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+///     returns true.
+///   * Any of the [RenderEditable]s it is attached to repaints.
+///   * The [notifyListeners] method is called, which typically happens when the
+///     painter's attributes change.
+///
+/// See also:
+///  * [RenderEditable.setForegroundPainter], which takes a [RenderEditablePainter]
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:2445]: nit: space before this line

File: packages/flutter/lib/src/rendering/editable.dart [hunk 21]
```diff
@@ -262,40 +265,135 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
        _forceLine = forceLine,
-       _promptRectRange = promptRectRange,
        _clipBehavior = clipBehavior {
     assert(_showCursor != null);
     assert(!_showCursor.value || cursorColor != null);
     this.hasFocus = hasFocus ?? false;
-    if (promptRectColor != null)
-      _promptRectPaint.color = promptRectColor;
+
+    _selectionPainter.highlightColor = selectionColor;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:290]: For my understanding, the plan is to move all these painters out of this class?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 22]
```diff
@@ -262,40 +265,135 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+
+  // Caret Painters:
+  // The floating painter. This painter paints the regular caret as well.
+  late final _FloatingCursorPainter _caretPainter = _FloatingCursorPainter(_onCaretChanged);
+
+  // Text Highlight painters:
+  final _TextHighlightPainter _selectionPainter = _TextHighlightPainter();
+  final _TextHighlightPainter _autocorrectHighlightPainter = _TextHighlightPainter();
+
+  _CompositeRenderEditablePainter get _builtInForegounrdPainters => _cachedBuiltInForegroundPainters ??= _createBuiltInForegroundPainters();
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:376]:
```suggestion
  _CompositeRenderEditablePainter get _builtInForegroundPainters => _cachedBuiltInForegroundPainters ??= _createBuiltInForegroundPainters();

```

Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:376]: Also, you should be able to just defines this as:

```dart
late final _CompositeRenderEditablePainter _builtInForegounrdPainters = _createBuiltInForegroundPainters();
```

With late it will only get instantiated on first access.

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/editable.dart:376]: Oh, never mind. I see below you need the option to clear this.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/editable.dart:226]: Would it make sense to keep `painter` and `foregroundPainter` an implementation detail at this step of the refactoring and not expose them via the API until we move to the next step where they are actually used and other arguments are removed? To keep the API surface from getting bigger...

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:226]: Yeah I can do that, but I think I have a few tests to ensure adding/changing external painters works. Should I keep those tests for the next PR?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 21 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:290]: Ideally, all of them.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 19 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:2438]: reindented to 1 space and added an empty line between the bullet points(I think you meant the latter?)

**Review by goderbauer — APPROVED** [review]
LGTM
